src/solvers/README.md

```diff
@@ -6,59 +6,155 @@
+ideas remain the same.
+
+
+\section solvers-interfaces Key Interfaces
+
+The most basic interface is `decision_proceduret` .  It gives the
+interface of all decision procedures.  You call `set_to_true` and
+`set_to_false` to give the formulae and then `dec_solve` to check if
+they are satisfiable.  If they are, it returns `D_SATISFIABLE` and you
+can use `get` to find the values in the satisfying assignment (if the
+underlying decision procedure supports this).  If you are implementing
+a solver, then this is the most basic interface you have to support,
+if you are using the solver, this is the best interface to use as it
+does not commit you to any particular kind of solver.  Looking at the
+inheritance diagram from `decision_proceduret` is a good way of
+getting an over-view of the solvers currently supported.
+
+Many (but not all) decision procedures have a notion of logical
+expression and can provide information about logical expressions
+within the solver.  `prop_convt` expands on the interface of
+`decision_proceduret` to add a data-type (`literalt`) and interfaces
+for manipulating logical expressions within the solver.
+
+Within decision procedures it is common to reduce the logical
+expressions to equivalent expressions in a simpler language.  This is
+similar to what a compiler will do in reducing higher-level language
+constructs to simple, assembler like instructions.  This, of course,
+relies on having a decision procedure for the simpler language, just
+as a compiler relies on you having a processor that can handle the
+assembler.  One of the popular choices of "processor" for decision
+procedures are SAT solvers.  These handle a very restricted language
+in which all variables are simple Booleans and all formula are just
+made of logical gates.  By keeping their input simple, they can be
+made very fast and efficient; kind of like RISC processors.  Like
+processors, creating a good SAT solver is a very specialised skill, so
+CPROVER uses third-party SAT solvers.  By default this is MiniSAT, but
+other are supported (see the `sat/` directory).  To do this it needs a
+software interface to a SAT solver : this is `propt`.  It uses the
+same `literalt` to refer to Boolean variables, just as `prop_convt`
+uses them to refer to logical expressions.  `land`, `lor`, `lxor` and
+so on allow gates to be constructed to express the formulae to be
+solved.  If `cnf_handled_well` is true then you may also use `lcnf` to
+build formulae.  Finally, `prop_solve` will run the decision procedure.
+
+As previously mentioned, many decision procedures reduce formulae to
+CNF and solve with a SAT solver.  `prop_conv_solvert` contains the
+foundations of this conversion.  It implements the `prop_conv` by
+having an instance of `propt` (a SAT solver) and reducing the
+expressions that are input into CNF.  The key entry point to this
+procedure is `convert` which then splits into `convert_boolean` (which
```

Comment:
prefix function names by the class name to make doxygen link it correctly

Reply:
Done; I think.